postgres/datadog_checks/postgres/explain_parameterized_queries.py

```diff
@@ -0,0 +1,181 @@
+
+        Note, prepared statements created by the user defined in the integration config are not persisted.
+        When the session ends, all prepared statements are automatically deallocated by Postgres.
+
+        Walkthrough:
+            1. Set the plan cache mode: `SET plan_cache_mode = force_generic_plan;`
+            2. Create a prepared statement: `PREPARE dd_products AS SELECT * FROM products WHERE id = $1;`
+            3. Query `pg_prepared_statements` to determine how many parameters a query requires
+            and provide generic values (null).
+            4. Execute and explain: `EXPLAIN EXECUTE dd_products(null);`
+                Returns: (plan)
+    '''
+
+    def __init__(self, check, config):
+        self._check = check
+        self._config = config
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def explain_statement(self, dbname, statement, obfuscated_statement):
+        if self._check.version < V12:
+            return None
+        self._set_plan_cache_mode(dbname)
+
+        query_signature = compute_sql_signature(obfuscated_statement)
+        if not self._create_prepared_statement(dbname, statement, obfuscated_statement, query_signature):
+            return None
+
+        result = self._explain_prepared_statement(dbname, statement, obfuscated_statement, query_signature)
+        self._deallocate_prepared_statement(dbname, query_signature)
+        if result:
+            return result[0][0][0]
+        return None
+
+    def _set_plan_cache_mode(self, dbname):
+        self._execute_query(dbname, "SET plan_cache_mode = force_generic_plan")
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def _create_prepared_statement(self, dbname, statement, obfuscated_statement, query_signature):
+        try:
+            self._execute_query(
+                dbname,
+                PREPARE_STATEMENT_QUERY.format(query_signature=query_signature, statement=statement),
+            )
+            return True
+        except Exception as e:
+            if self._config.log_unobfuscated_plans:
+                logger.warning(
+                    'Failed to create prepared statement when explaining statement(%s)=[%s] | err=[%s]',
+                    query_signature,
+                    statement,
+                    e,
+                )
+            else:
+                logger.warning(
+                    'Failed to create prepared statement when explaining statement(%s)=[%s] | err=[%s]',
+                    query_signature,
+                    obfuscated_statement,
+                    e,
+                )
+        return False
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def _get_number_of_parameters_for_prepared_statement(self, dbname, query_signature):
+        rows = self._execute_query_and_fetch_rows(
+            dbname, PARAM_TYPES_FOR_PREPARED_STATEMENT_QUERY.format(query_signature=query_signature)
+        )
```

Comment:
According to this: https://www.postgresql.org/docs/current/view-pg-prepared-statements.html `parameter_types` is an array of `regtype`. Can't you just use `array_length` function (see here: https://www.postgresql.org/docs/8.4/functions-array.html) to get the length inside the SQL query instead of doing string parsing?

Reply:
I see Zach has the same feedback.

Reply:
https://github.com/DataDog/integrations-core/pull/13434/commits/362742808dba37ada80011f880e3a3cf100abf47